### PR TITLE
[Experiment] Macro-based JsonSchema derivation

### PR DESCRIPTION
**Pull request by krzemin**
Following #144 and discussion on gitter, I implemented `JsonSchema` derivation with a macro.

I built more exhaustive tests checking support for few border cases (like coproduct with single branch or types with type parameters). Ideally I think they should be shared with `json-schema-generic` (shapeless based).

I tested this implementation on larger codebase and it seems we've got quite serious compilation time performance improvement (from 30s to under 5s).

Also, error reporting points exactly to the place where the problem is, like in this example:

```
[info] /Users/krzemin/Projects/OpenSource/endpoints/json-schema/json-schema-macros/src/test/scala/endpoints/macros/JsonSchemasTest.scala:60: Can't derive JsonSchema[Short] required for field baz!
[info]   case class Foo(bar: String, baz: Short, qux: Option[Boolean])
[info]                               ^
[info] /Users/krzemin/Projects/OpenSource/endpoints/json-schema/json-schema-macros/src/test/scala/endpoints/macros/JsonSchemasTest.scala:68: Can't derive JsonSchema[List[GenericSchemas.this.Foo]] required for field ss!
[info]   case class QuuxA(ss: List[Foo]) extends Quux
[info]                    ^
[info] /Users/krzemin/Projects/OpenSource/endpoints/json-schema/json-schema-macros/src/test/scala/endpoints/macros/JsonSchemasTest.scala:68: Can't derive JsonSchema[GenericSchemas.this.QuuxA] required for coproduct type GenericSchemas.this.Quux!
[info]   case class QuuxA(ss: List[Foo]) extends Quux
[info]              ^
[error] /Users/krzemin/Projects/OpenSource/endpoints/json-schema/json-schema-macros/src/test/scala/endpoints/macros/JsonSchemasTest.scala:93: Can't derive JsonSchema[GenericSchemas.this.Quux]
[error]     implicit val schema: JsonSchema[Quux] = genericJsonSchema[Quux]
[error]                                                              ^ 
```

### CODE REVIEW

File: openapi/openapi/src/test/scala/endpoints/openapi/ReferencedSchemaTest.scala [hunk 1]
```diff
@@ -1,17 +1,14 @@
...
 
 class ReferencedSchemaTest extends WordSpec with Matchers {
 
   sealed trait Storage
-
-  object Storage {
-    case class Library(room: String, shelf: Int) extends Storage
-    case class Online(link: String) extends Storage
-  }
+  case class StorageLibrary(room: String, shelf: Int) extends Storage
```
Comment (krzemin) [inline openapi/openapi/src/test/scala/endpoints/openapi/ReferencedSchemaTest.scala]: Having classes defined inside companion object caused `StackOverflowError ` in the compiler. It may be related to this bug: https://github.com/scala/bug/issues/11157

File: json-schema/json-schema-macros/src/main/scala/endpoints/macros/Macros.scala [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+  final def genericJsonSchemaImpl[A: c.WeakTypeTag]: c.Tree = {
+    val A = weakTypeOf[A]
+    findJsonSchema(A).getOrElse {
+      c.abort(c.enclosingPosition, s"Can't derive JsonSchema[$A]")
+    }
+  }
+
+  private def findJsonSchema(tpe: Type, allowImplicitSearch: Boolean = false): Option[c.Tree] = {
+    findPrimitive(tpe)
```
[the remark this answers is not in the document]
Comment (krzemin) [reply json-schema/json-schema-macros/src/main/scala/endpoints/macros/Macros.scala]: Well, they do - thanks for hint!

However in macros tests, where we have trivial string-based interpreter, compiler initially couldn't find primitive instances through implicit search due to ambiguities caused by having `type JsonSchema[A] = String`. It can be fixed by tagging the type in test interpreter (i.e. `type JsonSchema[A] = String with Tag[A]`). OTOH direct primitive approach worked well also for trivial interpreter. No strong opinion here which one is better :)

File: openapi/build.sbt [hunk 3]
```diff
@@ -8,7 +8,7 @@ lazy val openapi =
       name := "endpoints-openapi",
       libraryDependencies += "io.circe" %%% "circe-core" % circeVersion
     )
-    .dependsOnLocalCrossProjects("json-schema-generic")
+    .dependsOnLocalCrossProjects("json-schema-macros")
```
[the remark this answers is not in the document]
Comment (krzemin) [reply openapi/build.sbt]: Sure, it was only for the sake of testing on some non-trivial interpreter :)

File: json-schema/json-schema-macros/src/main/scala/endpoints/macros/Macros.scala [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+  private def findJsonSchema(tpe: Type, allowImplicitSearch: Boolean = false): Option[c.Tree] = {
+    {
+      if(primitiveTypes.exists(tpe <:< _) || allowImplicitSearch)
+        findImplicit(tq"JsonSchema[$tpe]")
+      else
+        None
+    }.orElse {
+        if(tpe <:< typeOf[Seq[_]]) {
+          findSeq(tpe)
```
[the remark this answers is not in the document]
Comment (krzemin) [reply json-schema/json-schema-macros/src/main/scala/endpoints/macros/Macros.scala]: Tried this, but stuck at the weird error that I can't understand. I would expect that during implicit  expansion of `arrayJsonSchema` for `JsonSchema[List[GenericSchemas.this.Foo]]`, as required implicit parameter of type `JsonSchema[GenericSchemas.this.Foo]` compiler picks `GenericSchemas.this.Foo.schema`, but for some reason it doesn't.

```
findJsonSchema(GenericSchemas.this.QuuxA)
findJsonSchema(List[GenericSchemas.this.Foo])
findImplicit(JsonSchema[List[GenericSchemas.this.Foo]])
[info] /Users/krzemin/Projects/OpenSource/endpoints/json-schema/json-schema-macros/src/test/scala/endpoints/macros/JsonSchemasTest.scala:69: arrayJsonSchema is not a valid implicit value for GenericSchemas.this.JsonSchema[GenericSchemas.this.Foo] because:
[info] hasMatchingSymbol reported error: polymorphic expression cannot be instantiated to expected type;
[info]  found   : [C[X] <: Seq[X], A]GenericSchemas.this.JsonSchema[C[A]]
[info]  required: GenericSchemas.this.JsonSchema[GenericSchemas.this.Foo]
[info]     implicit val schema: JsonSchema[QuuxA] = genericJsonSchema[QuuxA]
[info]                                                               ^
[info] /Users/krzemin/Projects/OpenSource/endpoints/json-schema/json-schema-macros/src/test/scala/endpoints/macros/JsonSchemasTest.scala:69: arrayJsonSchema is not a valid implicit value for GenericSchemas.this.JsonSchema[List[GenericSchemas.this.Foo]] because:
[info] hasMatchingSymbol reported error: could not find implicit value for parameter jsonSchema: GenericSchemas.this.JsonSchema[GenericSchemas.this.Foo]
[info]     implicit val schema: JsonSchema[QuuxA] = genericJsonSchema[QuuxA]
[info]                                                               ^
implicit found for JsonSchema[List[GenericSchemas.this.Foo]]: <empty>
[info] /Users/krzemin/Projects/OpenSource/endpoints/json-schema/json-schema-macros/src/test/scala/endpoints/macros/JsonSchemasTest.scala:67: Can't derive JsonSchema[List[GenericSchemas.this.Foo]] required for field ss!
[info]   case class QuuxA(ss: List[Foo]) extends Quux
[info]                    ^
[error] /Users/krzemin/Projects/OpenSource/endpoints/json-schema/json-schema-macros/src/test/scala/endpoints/macros/JsonSchemasTest.scala:69: Can't derive JsonSchema[GenericSchemas.this.QuuxA]
[error]     implicit val schema: JsonSchema[QuuxA] = genericJsonSchema[QuuxA]
[error]                                                               ^
```

Pushed the changes with few debug logs (with `-Xlog-implicits` enabled).

**Comment by guersam** [comment]
Although I'm not that familiar with macro, the compilation time reduction and improved error reporting sound beneficial.

If you want to add `@documentation` support by yourself please feel free to cherrypick the test cases from #128, or I'll take care of it after this PR is merged.

File: openapi/build.sbt (hunk 3 above)
[the remark this answers is not in the document]
Comment (krzemin) [reply openapi/build.sbt]: Tested again - openapi tests pass on both implementations. But reverted to generic, as suggested.

File: json-schema/json-schema-macros/src/main/scala/endpoints/macros/Macros.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (krzemin) [reply json-schema/json-schema-macros/src/main/scala/endpoints/macros/Macros.scala]: Finally had some time to investigate it a bit more.

- trying to derive the same instance from REPL made it work well (no code changed!)
- pulling `Foo` implicit json schema instance out of companion object helped as well

To me it seems like some weird compiler edge case, but hopefully it can be easily worked around.

I reorganized structure of example types and their json schemas in tests and it seems that instances for collections are now correctly picked by implicit resolution.

**Comment by krzemin** [comment]
Closing, as explained in #144.
